api/models/v0.3/modelBarcode.js

```diff
@@ -1,65 +1,170 @@
+      newBarcode = false;
+    } catch (error) {
+      // There are no unused barcodes so get the lowest barcode that was
+      // used and subtract one from it. This is the next new available one.
+      query =
+        "select barcode from barcodes where used=true order by barcodes asc limit 1;";
+      result = await db.query(query);
+      barcode = `${parseInt(result.rows[0].barcode, 10) - 1}`;
+      newBarcode = true;
+    }
 
-      this.markUsed(barcode.barcode);
+    return { barcode, newBarcode };
+  }
 
-      // # make sure barcode is available on ILS
-      barcodeFound = client.available(barcode.barcode);
+  /**
+   * release()
+   * Close the pool connection to the database.
+   */
+  async release() {
+    await db.release();
+  }
 
+  /**
+   * markUsed(barcode, used)
+   * Set a barcode to used or not used in the database.
+   * @param {string} barcode
+   * @param {boolean} used
+   */
+  async markUsed(barcode, used = false) {
+    const query = `UPDATE barcodes SET used=${used} WHERE barcode='${barcode}';`;
+    await db.query(query);
+  }
+
+  /**
+   * freeBarcode(barcode)
+   * Set an existing barcode to unused.
+   * @param {string} barcode
+   */
+  async freeBarcode(barcode) {
+    await this.markUsed(barcode, false);
+  }
+
+  /**
+   * addBarcode(barcode, used)
+   * Add a new barcode to the database and set used to true or false (default).
+   * @param {string} barcode
+   * @param {boolean} used
+   */
+  async addBarcode(barcode, used = false) {
+    const query = `INSERT INTO barcodes (barcode, used) VALUES ('${barcode}', ${used});`;
+    await db.query(query);
+  }
+
+  /**
+   * availableInIls(barcode, newBarcode, tries)
+   * Check the current barcode's availability in the ILS. It will try the
+   * next barcode in the sequence until an available one is found (based on
+   * the amount of tries). If the barcode is available, return it but also
+   * add it to the database and check it off as used. If the barcode was
+   * originally not new (so already in the database marked as unused), then
+   * just update the used value to true.
+   *
+   * @param {string} barcode
+   * @param {boolean} newBarcode
+   * @param {number} tries
+   */
+  async availableInIls(barcode, newBarcode, tries = 5) {
+    const isBarcode = true;
+    let barcodeAvailable = false;
+    let barcodeToTry = barcode;
+
+    while (!barcodeAvailable && tries > 0) {
+      // make sure barcode is available on ILS
+      barcodeAvailable = await this.ilsClient.available(
+        barcodeToTry,
+        isBarcode
+      );
+
+      // If the barcode is available, insert it into the database.
+      if (barcodeAvailable) {
+        if (newBarcode) {
+          this.addBarcode(barcodeToTry, true);
+        } else {
+          this.markUsed(barcodeToTry, true);
+        }
+        break;
+      }
+
+      // Otherwise, let's try once more with the next barcode in the sequence.
       tries -= 1;
+      barcodeToTry = `${barcodeToTry - 1}`;
```

Comment:
Antonio mentioned in the call yesterday that it's more complicated than this because the last digit of a barcode is a checksum digit.

So if the last used barcode is `233331234567895`, the first available barcode is not `233331234567894`, it's `23333123456788X`, where X is the result of the checksum algorithm.

This requirement went unstated last time, and it's caused a lot of problems, so I'm making a big deal out of it now. This should be a separate function. It's OK to implement the function as x-1 for now, but once you have the checksum algorithm it should implement that algorithm.

Reply:
Yeah, I just got the algorithm this morning and I'll implement it soon.